src/software/new_geom/segment.cpp

```diff
@@ -0,0 +1,240 @@
+        // coordinates, choose x because we know there is variance in these values
+        return (point.x() <= getStart().x() && point.x() >= getEnd().x()) ||
+               (point.x() <= getEnd().x() && point.x() >= getStart().x());
+    }
+
+    return false;
+}
+
+bool Segment::collinear(const Segment& other, double fixed_epsilon,
+                        int ulps_distance) const
+{
+    // Two segments are collinear if all Points are collinear
+    if (Point::collinear(getStart(), getEnd(), other.getStart(), fixed_epsilon,
+                         ulps_distance) &&
+        Point::collinear(getStart(), getEnd(), other.getEnd(), fixed_epsilon,
+                         ulps_distance))
+    {
+        return true;
+    }
+    return false;
+}
+
+bool Segment::intersects(const Segment& other) const
+{
+    // Using the FASTER LINE SEGMENT INTERSECTION algorithm from p.199 of Graphics Gems
+    // III (IBM Version)
+    // https://www.sciencedirect.com/science/article/pii/B9780080507552500452.
+
+    // Values are pre-computed to improve performance
+    const double p1x(getStart().x());
+    const double p1y(getStart().y());
+    const double p2x(getEnd().x());
+    const double p2y(getEnd().y());
+    const double p3x(other.getEnd().x());
+    const double p3y(other.getEnd().y());
+    const double p4x(other.getStart().x());
+    const double p4y(other.getStart().y());
+
+    const double ax = p2x - p1x;
+    const double ay = p2y - p1y;
+    const double bx = p3x - p4x;
+    const double by = p3y - p4y;
+    const double cx = p1x - p3x;
+    const double cy = p1y - p3y;
+
+    double denominator = ay * bx - ax * by;
+    double numerator1  = by * cx - bx * cy;
+    if (denominator > 0)
+    {
+        if (numerator1 < 0 || numerator1 > denominator)
+        {
+            return false;
+        }
+    }
+    else
+    {
+        if (numerator1 > 0 || numerator1 < denominator)
+        {
+            return false;
+        }
+    }
+
+    // Only compute numerator2 once we're sure we need it
+    double numerator2 = ax * cy - ay * cx;
+    if (denominator > 0)
+    {
+        if (numerator2 < 0 || numerator2 > denominator)
+        {
+            return false;
+        }
+    }
+    else
+    {
+        if (numerator2 > 0 || numerator2 < denominator)
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+std::vector<Point> Segment::intersection(const Segment& other) const
+{
+    std::vector<Point> output;
+
+    Point a = getStart();
+    Point b = getEnd();
+    Point c = other.getStart();
+    Point d = other.getEnd();
+
+    // check for overlaps
+    if (other.contains(a))
+    {
+        output.emplace_back(a);
+    }
+    if (other.contains(b))
+    {
+        output.emplace_back(b);
+    }
+    if (contains(c))
+    {
+        output.emplace_back(c);
+    }
+    if (contains(d))
+    {
+        output.emplace_back(d);
+    }
```

Comment:
What would happen if both segments were identical? The spec says we only return 2 points, worried in the case that they are equal, we might return 4 points?

Reply:
I will add a test

Reply:
fixed a bug good catch